File: web/src/lib/api/pod.ts
import { request, orgPath } from "./base";
import type { PodMode } from "@/lib/pod-modes";

// Pod interface matching the store
export interface PodData {
  id: number;
  pod_key: string;
  status: "initializing" | "running" | "paused" | "disconnected" | "orphaned" | "completed" | "terminated" | "error" | "failed";
  agent_status: string;
  prompt?: string;
  branch_name?: string;
  sandbox_path?: string;
  started_at?: string;
  finished_at?: string;
  last_activity?: string;
  created_at: string;
  title?: string; // OSC 0/2 terminal title
  alias?: string; // User-defined display name
  runner?: {
    id: number;
    node_id: string;
    status: string;
  };
  agent?: {
    name: string;
    slug: string;
  };
  repository?: {
    id: number;
    name: string;
    slug: string;
    provider_type?: string; // github, gitlab, gitee
  };
  ticket?: {
    id: number;
    slug: string;
    title: string;
  };
  loop?: {
    id: number;
    name: string;
    slug: string;
  };
  interaction_mode?: PodMode;
  perpetual?: boolean;
  restart_count?: number;
  last_restart_at?: string;
  error_code?: string;
  error_message?: string;
  created_by?: {
    id: number;
    username: string;
    name?: string;
  };
}

// Pods API
export const podApi = {
  list: (filters?: { status?: string; runnerId?: number; createdById?: number; limit?: number; offset?: number }) => {
    const params = new URLSearchParams();
    if (filters?.status) params.append("status", filters.status);
    if (filters?.runnerId) params.append("runner_id", String(filters.runnerId));
    if (filters?.createdById) params.append("created_by_id", String(filters.createdById));
    if (filters?.limit) params.append("limit", String(filters.limit));
    if (filters?.offset) params.append("offset", String(filters.offset));
    const query = params.toString() ? `?${params.toString()}` : "";
    return request<{ pods: PodData[]; total: number; limit: number; offset: number }>(`${orgPath("/pods")}${query}`);
  },

  get: (key: string) =>
    request<{ pod: PodData }>(`${orgPath("/pods")}/${key}`),

  create: (data: {
    agent_slug?: string; // Required unless resuming
    runner_id?: number;
    ticket_slug?: string;
    alias?: string; // User-defined display name (max 100 chars)
    cols?: number; // Terminal columns (from xterm.js)
    rows?: number; // Terminal rows (from xterm.js)
    // AgentFile Layer — SSOT (PROMPT, MODE, CONFIG, REPO, BRANCH, CREDENTIAL)
    agentfile_layer?: string;
    // Resume mode fields
    source_pod_key?: string;
    resume_agent_session?: boolean;
    // Perpetual mode
    perpetual?: boolean;
  }) =>
    request<{ message: string; pod: PodData }>(
      orgPath("/pods"),
      {
        method: "POST",
        body: data,
      }
    ),

  terminate: (key: string) =>
    request<{ message: string }>(`${orgPath("/pods")}/${key}/terminate`, {
      method: "POST",
    }),

  // Get connection info for WebSocket terminal
  getConnectionInfo: (key: string) =>
    request<{ pod_key: string; ws_url: string; status: string }>(
      `${orgPath("/pods")}/${key}/connect`
    ),

  // Get Pod connection info via Relay
  // Returns Relay URL and token for WebSocket connection
  // Note: podKey is embedded in the token for channel routing
  getPodConnection: (key: string) =>
    request<{
      relay_url: string;
      token: string;
      pod_key: string;
    }>(`${orgPath("/pods")}/${key}/relay/connect`),

  // Update pod alias (user-defined display name)
  // Pass null to clear the alias
  updateAlias: (key: string, alias: string | null) =>
    request<{ message: string }>(`${orgPath("/pods")}/${key}/alias`, {
      method: "PATCH",
      body: { alias },
    }),
};

File: clients/core/crates/ffi/src/dto/pod.rs
use agentsmesh_types::{
    CreatePodRequest, Pod, PodAgentInfo, PodConnectionInfo, PodCreatedByInfo, PodListResponse,
    PodLoopInfo, PodRepositoryInfo, PodRunnerInfo, PodStatus, PodTicketInfo, UpdatePodAliasRequest,
};

#[derive(Clone, Copy, Debug, uniffi::Enum)]
pub enum PodStatusDto {
    Pending,
    Creating,
    Initializing,
    Running,
    Paused,
    Stopping,
    Disconnected,
    Orphaned,
    Completed,
    Terminated,
    Error,
    Failed,
    Unknown,
}

impl From<PodStatus> for PodStatusDto {
    fn from(s: PodStatus) -> Self {
        match s {
            PodStatus::Pending => Self::Pending,
            PodStatus::Creating => Self::Creating,
            PodStatus::Initializing => Self::Initializing,
            PodStatus::Running => Self::Running,
            PodStatus::Paused => Self::Paused,
            PodStatus::Stopping => Self::Stopping,
            PodStatus::Disconnected => Self::Disconnected,
            PodStatus::Orphaned => Self::Orphaned,
            PodStatus::Completed => Self::Completed,
            PodStatus::Terminated => Self::Terminated,
            PodStatus::Error => Self::Error,
            PodStatus::Failed => Self::Failed,
            PodStatus::Unknown => Self::Unknown,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodRunnerInfoDto {
    pub id: Option<i64>,
    pub node_id: Option<String>,
    pub status: Option<String>,
}

impl From<PodRunnerInfo> for PodRunnerInfoDto {
    fn from(r: PodRunnerInfo) -> Self {
        Self {
            id: r.id,
            node_id: r.node_id,
            status: r.status,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodAgentInfoDto {
    pub name: Option<String>,
    pub slug: Option<String>,
}

impl From<PodAgentInfo> for PodAgentInfoDto {
    fn from(a: PodAgentInfo) -> Self {
        Self {
            name: a.name,
            slug: a.slug,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodRepositoryInfoDto {
    pub id: Option<i64>,
    pub name: Option<String>,
    pub slug: Option<String>,
    pub provider_type: Option<String>,
}

impl From<PodRepositoryInfo> for PodRepositoryInfoDto {
    fn from(r: PodRepositoryInfo) -> Self {
        Self {
            id: r.id,
            name: r.name,
            slug: r.slug,
            provider_type: r.provider_type,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodTicketInfoDto {
    pub id: Option<i64>,
    pub slug: Option<String>,
    pub title: Option<String>,
}

impl From<PodTicketInfo> for PodTicketInfoDto {
    fn from(t: PodTicketInfo) -> Self {
        Self {
            id: t.id,
            slug: t.slug,
            title: t.title,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodLoopInfoDto {
    pub id: Option<i64>,
    pub name: Option<String>,
    pub slug: Option<String>,
}

impl From<PodLoopInfo> for PodLoopInfoDto {
    fn from(l: PodLoopInfo) -> Self {
        Self {
            id: l.id,
            name: l.name,
            slug: l.slug,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodCreatedByInfoDto {
    pub id: Option<i64>,
    pub username: Option<String>,
    pub name: Option<String>,
}

impl From<PodCreatedByInfo> for PodCreatedByInfoDto {
    fn from(u: PodCreatedByInfo) -> Self {
        Self {
            id: u.id,
            username: u.username,
            name: u.name,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodDto {
    pub key: String,
    pub id: Option<i64>,
    pub status: PodStatusDto,
    pub agent_status: Option<String>,
    pub alias: Option<String>,
    pub title: Option<String>,
    pub agent_slug: String,
    pub runner_id: Option<i64>,
    pub runner_name: Option<String>,
    pub user_id: Option<i64>,
    pub ticket_slug: Option<String>,
    pub channel_id: Option<i64>,
    pub runner: Option<PodRunnerInfoDto>,
    pub agent: Option<PodAgentInfoDto>,
    pub repository: Option<PodRepositoryInfoDto>,
    pub ticket: Option<PodTicketInfoDto>,
    pub loop_info: Option<PodLoopInfoDto>,
    pub created_by: Option<PodCreatedByInfoDto>,
    pub prompt: Option<String>,
    pub branch_name: Option<String>,
    pub sandbox_path: Option<String>,
    pub started_at: Option<String>,
    pub finished_at: Option<String>,
    pub last_activity: Option<String>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub interaction_mode: Option<String>,
    pub perpetual: Option<bool>,
    pub restart_count: Option<i32>,
    pub last_restart_at: Option<String>,
    pub error_code: Option<String>,
    pub error_message: Option<String>,
}

impl From<Pod> for PodDto {
    fn from(p: Pod) -> Self {
        Self {
            key: p.key,
            id: p.id,
            status: p.status.into(),
            agent_status: p.agent_status,
            alias: p.alias,
            title: p.title,
            agent_slug: p.agent_slug,
            runner_id: p.runner_id,
            runner_name: p.runner_name,
            user_id: p.user_id,
            ticket_slug: p.ticket_slug,
            channel_id: p.channel_id,
            runner: p.runner.map(Into::into),
            agent: p.agent.map(Into::into),
            repository: p.repository.map(Into::into),
            ticket: p.ticket.map(Into::into),
            loop_info: p.loop_info.map(Into::into),
            created_by: p.created_by.map(Into::into),
            prompt: p.prompt,
            branch_name: p.branch_name,
            sandbox_path: p.sandbox_path,
            started_at: p.started_at,
            finished_at: p.finished_at,
            last_activity: p.last_activity,
            created_at: p.created_at,
            updated_at: p.updated_at,
            interaction_mode: p.interaction_mode,
            perpetual: p.perpetual,
            restart_count: p.restart_count,
            last_restart_at: p.last_restart_at,
            error_code: p.error_code,
            error_message: p.error_message,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodConnectionInfoDto {
    pub relay_url: String,
    pub token: String,
    pub pod_key: String,
}

impl From<PodConnectionInfo> for PodConnectionInfoDto {
    fn from(i: PodConnectionInfo) -> Self {
        Self {
            relay_url: i.relay_url,
            token: i.token,
            pod_key: i.pod_key,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct CreatePodRequestDto {
    pub agent_slug: String,
    pub agentfile_layer: Option<String>,
    pub runner_id: Option<i64>,
    pub alias: Option<String>,
    pub ticket_slug: Option<String>,
    pub cols: Option<u16>,
    pub rows: Option<u16>,
    pub source_pod_key: Option<String>,
    pub resume_agent_session: Option<bool>,
    pub perpetual: Option<bool>,
}

impl From<CreatePodRequestDto> for CreatePodRequest {
    fn from(d: CreatePodRequestDto) -> Self {
        Self {
            agent_slug: d.agent_slug,
            agentfile_layer: d.agentfile_layer,
            runner_id: d.runner_id,
            alias: d.alias,
            ticket_slug: d.ticket_slug,
            cols: d.cols,
            rows: d.rows,
            source_pod_key: d.source_pod_key,
            resume_agent_session: d.resume_agent_session,
            perpetual: d.perpetual,
        }
    }
}

#[derive(Clone, Debug, uniffi::Record)]
pub struct PodListResponseDto {
    pub pods: Vec<PodDto>,
    pub total: Option<i64>,
}

impl From<PodListResponse> for PodListResponseDto {
    fn from(r: PodListResponse) -> Self {
        Self {
            pods: r.pods.into_iter().map(PodDto::from).collect(),
            total: r.total,
        }
    }
}

/// Bridge Swift's `String` alias back to the strong-typed request shape.
pub(crate) fn update_pod_alias_req(alias: String) -> UpdatePodAliasRequest {
    UpdatePodAliasRequest { alias }
}
